### PR TITLE
Flatpak

### PR DESCRIPTION
**Pull request by jmercouris**
I think it works :-O

can someone please try the resultant Flatpak? I tried it myself, but I want to make sure it isn't because I had installed something to make it work!

### CODE REVIEW

**Comment by aadcg** [comment]
@jmercouris Thanks for working on this!

For those reviewing this PR that are on Guix. `flatpak` is available, while `flatpak-builder` isn't. I recommend installing the latter via `nix` (add `(service nix-service-type)` to your Guix services). Theoretically, `flatpak` could be installed via `nix` too, but if you're using `nix` as an addition package manager (not NixOS) [you're out of luck](https://nixos.wiki/wiki/Flatpak). 

If the flatpak commands won't work, append `--user` so that they're executed user-wide (instead of system-wide).

`flatpak remote-add --user --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo`
`flatpak install --user --noninteractive org.gnome.Sdk/x86_64/41`

**Comment by aadcg** [comment]
@jmercouris I was able to run the Nyxt flatpak!

File: makefile [hunk 1]
```diff
@@ -89,3 +76,21 @@ clean-submodules:
...
 clean: clean-submodules
+	rm -rf build
+
+.PHONY: flatpak-build
+flatpak-build:
+	@$(FLATPAK_BUILDER) --user --install --force-clean build $(FLATPAK_MANIFEST)
+
+.PHONY: flatpak-run
+flatpak-run:
+	@$(FLATPAK_COMMAND) run $(FLATPAK_APP_ID)
```
Comment (aadcg) [inline makefile:87]: Personally, I think `flatpak-run` doesn't belong to a makefile!

File: makefile [hunk 2]
```diff
@@ -89,3 +76,21 @@ clean-submodules:
...
+
+.PHONY: flatpak-repository
+flatpak-repository:
+	mkdir -p $(FLATPAK_EXPORT_REPOSITORY)
+	@$(FLATPAK_BUILDER) --force-clean --repo=$(FLATPAK_EXPORT_REPOSITORY) build $(FLATPAK_MANIFEST)
+
+.PHONY: flatpak-bundle
+flatpak-bundle:
+	@$(FLATPAK_COMMAND) build-bundle $(FLATPAK_EXPORT_REPOSITORY) nyxt.flatpak $(FLATPAK_APP_ID)
```
Comment (aadcg) [inline makefile:96]: Do we need all of these: `flatpak-repository`, `flatpak-bundle` and `flatpak-build`? What's the difference between these?

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -94,4 +94,4 @@ jobs:
       uses: ncipollo/release-action@v1
       with:
         bodyFile: release.txt
-        artifacts: "nyxt-*.tar.xz*,guix-recipe.txt,*.deb"
+        artifacts: "nyxt-*.tar.xz*,guix-recipe.txt,*.flatpak"
```
Comment (aadcg) [inline .github/workflows/release.yaml:97]: What exactly is this generated flatpak file?

**Comment by aadcg** [comment]
I'm confused about how we'll distribute the Nyxt flatpak.

There are 3 ways to install flatpaks. 

1. Via `flathub` ([we need to add the Nyxt flatpak to their repositories](https://github.com/flathub/flathub))
2. Via a so-called `.flatpakref` file that enables users to say `flatpak install /path/to/.flatpakref`.
3. Via a URL that points to the file mentioned above `flatpak install url-link-to-flatpakref`.

File: engineer.atlas.Nyxt.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
+app-id: engineer.atlas.Nyxt
+# Not a mistake! We need the sonames of shared libraries for SBCL during runtime.
+runtime: org.gnome.Sdk
+runtime-version: "41"
+sdk: org.gnome.Sdk
```
Comment (aadcg) [inline engineer.atlas.Nyxt.yaml:5]:
```log
Info: org.gnome.Sdk is end-of-life, with reason: The GNOME 41 runtime is no longer supported as of September 17, 2022. Please ask your application developer to migrate to a supported platform.
Info: org.gnome.Sdk.Locale is end-of-life, with reason: The GNOME 41 runtime is no longer supported as of September 17, 2022. Please ask your application developer to migrate to a supported platform.
Installing runtime/org.gnome.Sdk/x86_64/41
```

Do we really need this version? If so, can we add a comment explaining it?

**Comment by jmercouris** [comment]
> @jmercouris I was able to run the Nyxt flatpak!

Yay!

File: engineer.atlas.Nyxt.yaml (hunk 4 above)
Comment (jmercouris) [reply engineer.atlas.Nyxt.yaml:4]: You're right. I have updated it to 42! 43 is causing some gobject errors, so we will leave that for now.

File: .github/workflows/release.yaml (hunk 3 above)
Comment (jmercouris) [reply .github/workflows/release.yaml:97]: > I'm confused about how we'll distribute the Nyxt flatpak.
> 
> There are 3 ways to install flatpaks.
> 
>     1. Via `flathub` ([we need to add the Nyxt flatpak to their repositories](https://github.com/flathub/flathub))
> 
>     2. Via a so-called `.flatpakref` file that enables users to say `flatpak install /path/to/.flatpakref`.
> 
>     3. Via a URL that points to the file mentioned above `flatpak install url-link-to-flatpakref`.

We will support installation via two methods:

1. Flathub
2. A file called `nyxt.flatpak` which contains everything the user needs

what they do with this file is the following: `flatpak install nyxt.flatpak` and then it is available on their system :-)

File: makefile (hunk 2 above)
Comment (jmercouris) [reply makefile:96]: Yes we do need them all. In order to make a `flatpak-bundle` (it generates the standalone nyxt.flatpak file), we need a `flatpak-repository.` In order to test and run our flatpak locally we need `flatpak-build` and `flatpak-run`

File: makefile (hunk 1 above)
Comment (jmercouris) [reply makefile:87]: It makes a little bit easier than writing the tedious: `flatpak run engineer.atlas.Nyxt`, I've seen many other flatpak repositories that use this.

File: makefile (hunk 2 above)
Comment (aadcg) [reply makefile:96]: Clear.

File: .github/workflows/release.yaml (hunk 3 above)
Comment (aadcg) [reply .github/workflows/release.yaml:97]: Clear - the CI will make `nyxt.flatpak` available on each release.

File: engineer.atlas.Nyxt.yaml (hunk 4 above)
Comment (aadcg) [reply engineer.atlas.Nyxt.yaml:4]: Thanks.

**Comment by aadcg** [comment]
@jmercouris same comment: please merge according to our standards and squash some commits. 

For instance, having a commit where you specify version 41 and then 42 is useless. No one wants to see your thinking process while working on something. All we care about is: before there was no flatpak generation logic, and now there is. We want to see a clean addition of this feature. On the other hand, in the future, that generation logic will be tweaked. In that case, yes, I want to see every little detail because it has already landed on master!

**Comment by jmercouris** [comment]
@Ambrevar the files are still in the Git history. I removed them because I also delete code mostly instead of just commenting out. It is not gone, and since we are not using it, I think the clutter does not serve us well.

The work is not destroyed! It is still there! Do you disagree with my approach? Can you think of a better way I should have done it?

**Comment by aartaka** [comment]
> @Ambrevar the files are still in the Git history. I removed them because I also delete code mostly instead of just commenting out. It is not gone, and since we are not using it, I think the clutter does not serve us well.

GIt history tends to be unreliable, as it can be changed any moment by a recless rebase. We have a convention to not do this, but we are humans, which means we happen to err. 

I personally am in favor of keeping the .deb, even though it's not working for all Ubuntu versions. @jmercouris, you keep advocating for installability of Nyxt, why remove yet another way to install it then?

**Comment by Ambrevar** [comment]
@jmercouris There is no clutter if we keep the deb generation in a seperate file (I agree that keeping it commented would be ugly).

**Comment by jmercouris** [comment]
> @jmercouris There is no clutter if we keep the deb generation in a seperate file (I agree that keeping it commented would be ugly).

I still think it would be clutter, but that is not my rationale for removing it.

As we had mentioned in the other issue, it will not work on newer versions of Ubuntu, and it will confuse users. If we have a single way of installing (other than Guix), I think we can more easily support it. Thoughts?

**Comment by jmercouris** [comment]
> GIt history tends to be unreliable, as it can be changed any moment by a recless rebase. We have a convention to not do this, but we are humans, which means we happen to err.

I have thought about this argument, but a reckless rebase can destroy it even IF it is a file. Therefore, I don't believe there is a real difference between the two. If we happen to do such a rebase, well, it is OK, there are multiple copies of the Git history, one of us is bound to have it :-D 

> I personally am in favor of keeping the .deb, even though it's not working for all Ubuntu versions. @jmercouris, you keep advocating for installability of Nyxt, why remove yet another way to install it then?

Because we want installations that work, and the deb installation does not work, for a large quantity of the users. Many users on systems that do not support the deb will see it, and attempt to install it, then it won't work, and they will get frustrated. The flatpak always works, every time. Therefore, I think it is a better solution for our casual user. Plus, it is built into many Linux distributions! https://www.flatpak.org/setup/ I spent a lot of time evaluating the different casual user packaging options before narrowing down on Flatpak, which I believe to be the most promising.

**Comment by aartaka** [comment]
> > GIt history tends to be unreliable, as it can be changed any moment by a recless rebase. We have a convention to not do this, but we are humans, which means we happen to err.
> 
> I have thought about this argument, but a reckless rebase can destroy it even IF it is a file. Therefore, I don't believe there is a real difference between the two. 

There is. Getting really metaphysical: it's much easier to destroy a things that _was_, rather than the thing that _is_. It's harder to destroy both a file and a history of it, than just erasing history :P

> > I personally am in favor of keeping the .deb, even though it's not working for all Ubuntu versions. @jmercouris, you keep advocating for installability of Nyxt, why remove yet another way to install it then?
> 
> Because we want installations that work, and the deb installation does not work, for a large quantity of the users. Many users on systems that do not support the deb will see it, and attempt to install it, then it won't work, and they will get frustrated. The flatpak always works, every time. Therefore, I think it is a better solution for our casual user. Plus, it is built into many Linux distributions! https://www.flatpak.org/setup/ I spent a lot of time evaluating the different casual user packaging options before narrowing down on Flatpak, which I believe to be the most promising.

This is a valid argument, but [not for Ubuntu 23.04+ (The Register newspiece)](https://www.theregister.com/2023/02/23/ubuntu_remixes_drop_flatpak/)

**Comment by Ambrevar** [comment]
> As we had mentioned in the other issue, it will not work on newer versions of Ubuntu, and it will confuse users. If we have a single way of installing (other than Guix), I think we can more easily support it. Thoughts?

You just said it, we don't have a single way :p  Also we can always follow Ubuntu LTS, which some times will work for all latest versions of Ubuntu.  It's a matter of "Ubuntu period".

My point is not about that though, it's just about keeping this precious piece of code in the worktree.

**Comment by jmercouris** [comment]
Don't worry, it is not lost, it is still in the Git history!

I fundamentally disagree about keeping a broken installation offering in our downloads.

**Comment by Ambrevar** [comment]
It's not broken: if we want to release for the latest Ubuntu, it's a trivial change in the OS selection.

Hard work has been put into this, and the result is fantastic: a tailored .deb package that just works.  Please, let's be a bit more considerate here :p 

Being in the Git history is not a way to store code in a way that anyone  can realistically   find it back.

**Pull request merged**